Raise an anomaly when looking up unknown constant/inductive

SkySkimmer opened the pull request:
If you have access to a kernel name you also should have the
environment in which it is defined, barring hacks. In order to
disfavor hacks we make the standard lookups raise anomalies so that
people are forced to admit they rely on the internals of the
environment.

We find that hackers operated on the code for side effects, for
finding inductive schemes, for simpl and for Print Assumptions. They
attempted to operate on funind but the error handling code they wrote
would have raised another Not_found instead of being useful.

All these uses are indeed hacky so I am satisfied that we are not
forcing new hacks on callers.

SkySkimmer commented:
Bench started https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/726/console

SkySkimmer commented:
Bench is somewhat bad
~~~
┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬──────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │        mem faults        │
│                        │                         │                                             │                                             │                               │                          │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │     NEW    OLD   PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│             coq-sf-plf │   43.60   43.67 -0.16 % │      122837465403      122693085319 +0.12 % │      158410608148      158308582899 +0.06 % │     481480     488916 -1.52 % │      31      0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│          coq-fourcolor │ 2177.34 2180.31 -0.14 % │     6071717905493     6081362131339 -0.16 % │    11364714574152    11363249987851 +0.01 % │     929284     929252 +0.00 % │       1      0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│            coq-bignums │   69.43   69.40 +0.04 % │      193416484936      192990287481 +0.22 % │      253849437015      253684253499 +0.07 % │     491184     491452 -0.05 % │     188    191   -1.57 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│         coq-coquelicot │   75.72   75.60 +0.16 % │      208043575387      206864431860 +0.57 % │      250617975246      250453675131 +0.07 % │     656884     656784 +0.02 % │     214     11 +1845.45 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│   coq-mathcomp-algebra │  167.01  166.52 +0.29 % │      465770479163      464294428586 +0.32 % │      574522283980      573941045501 +0.10 % │     618200     618428 -0.04 % │      11      2 +450.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│  coq-mathcomp-fingroup │   51.92   51.75 +0.33 % │      144688622135      143912837186 +0.54 % │      187166589535      186969828407 +0.11 % │     567464     568028 -0.10 % │      85      1 +8400.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│              coq-color │  699.86  697.24 +0.38 % │     1958031416569     1949675983429 +0.43 % │     2327073215161     2325635516998 +0.06 % │    1353804    1382716 -2.09 % │     175    242  -27.69 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│         coq-verdi-raft │ 1330.95 1325.86 +0.38 % │     3712768051388     3697630222442 +0.41 % │     5004823952972     5001239500638 +0.07 % │    1822248    1822348 -0.01 % │       0      8 -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│     coq-mathcomp-field │  231.71  230.80 +0.39 % │      645678763928      643146995804 +0.39 % │      918999171188      918295803322 +0.08 % │     756072     756220 -0.02 % │       1      0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│              coq-verdi │  119.60  119.08 +0.44 % │      332354737496      330601677938 +0.53 % │      421766813036      421387880189 +0.09 % │     572236     572240 -0.00 % │      12      8  +50.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│              coq-flocq │  463.08  461.06 +0.44 % │     1288499902926     1282413821020 +0.47 % │     1777559171320     1776719416686 +0.05 % │    1293968    1293912 +0.00 % │     233     29 +703.45 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│             coq-geocoq │ 1948.86 1940.33 +0.44 % │     5439821855122     5409706611499 +0.56 % │     7845832587221     7842361302699 +0.04 % │    1251696    1251548 +0.01 % │     121    177  -31.64 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│            coq-unimath │ 2606.76 2595.04 +0.45 % │     7260204922362     7231264709506 +0.40 % │    13007455025844    13000052822443 +0.06 % │    1604740    1603744 +0.06 % │     490    179 +173.74 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-character │  188.82  187.70 +0.60 % │      525892230411      522971248721 +0.56 % │      694209502742      693424760686 +0.11 % │    1102352    1102176 +0.02 % │      14    104  -86.54 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│       coq-math-classes │  217.30  215.91 +0.64 % │      607325281691      604248499817 +0.51 % │      767979080365      767734303562 +0.03 % │     514876     514640 +0.05 % │      31     16  +93.75 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│       coq-fiat-parsers │  639.08  634.82 +0.67 % │     1787201635701     1777481746940 +0.55 % │     2690635886832     2688976308448 +0.06 % │    3233496    3242060 -0.26 % │     408    666  -38.74 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│          coq-fiat-core │  108.68  107.91 +0.71 % │      309192265840      308658178845 +0.17 % │      385523186491      385400143777 +0.03 % │     485568     485504 +0.01 % │     414    260  +59.23 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│               coq-hott │  335.46  333.08 +0.71 % │      910895736060      906340115250 +0.50 % │     1376873958770     1376610244837 +0.02 % │     477980     477900 +0.02 % │     448    109 +311.01 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│        coq-fiat-crypto │ 3808.69 3780.49 +0.75 % │    10581003226398    10504165377995 +0.73 % │    16525732191315    16516216116620 +0.06 % │    2582592    2521176 +2.44 % │    1035    721  +43.55 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│  coq-mathcomp-solvable │  191.24  189.69 +0.82 % │      532470052401      529420095833 +0.58 % │      708045954955      707159892395 +0.13 % │     774916     817792 -5.24 % │       4      1 +300.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-fiat-crypto-legacy │ 6247.66 6196.72 +0.82 % │    17397299132637    17253555670331 +0.83 % │    28151551886295    28084961592682 +0.24 % │    2662236    2585476 +2.97 % │     962    983   -2.14 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│        coq-lambda-rust │ 1618.74 1603.96 +0.92 % │     4510878341714     4472117954287 +0.87 % │     5955773175782     5951954002282 +0.06 % │    1644692    1569928 +4.76 % │      46      0    +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-ssreflect │   40.26   39.79 +1.18 % │      110844851833      110492750411 +0.32 % │      134090242358      134000354716 +0.07 % │     528592     528612 -0.00 % │      12     58  -79.31 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│               coq-corn │ 1471.91 1451.69 +1.39 % │     4107090357124     4051309412792 +1.38 % │     6024423242278     6015721239612 +0.14 % │     894928     894584 +0.04 % │       1      1   +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│ coq-mathcomp-odd-order │ 1481.40 1460.41 +1.44 % │     4129440882253     4070750910789 +1.44 % │     6978854400489     6974738242299 +0.06 % │    1267524    1267504 +0.00 % │     150     22 +581.82 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼──────────────────────────┤
│           coq-compcert │  974.47  957.18 +1.81 % │     2713781187189     2667418203543 +1.74 % │     3991259750073     3989297448310 +0.05 % │    1032860    1032820 +0.00 % │     237    253   -6.32 % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴──────────────────────────┘
~~~

mattam82 commented:
Might that be the ˋtry ... with` blocks alone?

SkySkimmer commented:
>Might that be the ˋtry ... with` blocks alone?

What else could it be?

mattam82 commented:
Would it be reasonable to provide fast path unsafe_lookup_* for tight loops using lookups knowing they’re well-scoped then?
I’m not really a fan of doing EnvMap.find in those cases

ppedrot commented:
Instead, I think we should export a `get` function in the Map API that fails an assertion when the key is not present. This makes the intent clearer and doesn't require catching the exception.

SkySkimmer commented:
>Instead, I think we should export a get function in the Map API that fails an assertion when the key is not present. This makes the intent clearer and doesn't require catching the exception.

We already expose one. I changed lookup_foo to use it and changed the implem to avoid try with not_found. Let's rebench.

ppedrot commented:
The `get` function is still implemented through `Not_found` catching though. Instead, it should be implemented by hand.

SkySkimmer commented:
>The get function is still implemented through Not_found catching though. Instead, it should be implemented by hand.

See commit...

SkySkimmer commented:
Bench looks ok
~~~

┌────────────────────────┬─────────────────────────┬─────────────────────────────────────────────┬─────────────────────────────────────────────┬───────────────────────────────┬─────────────────────────┐
│                        │      user time [s]      │                 CPU cycles                  │              CPU instructions               │     max resident mem [KB]     │       mem faults        │
│                        │                         │                                             │                                             │                               │                         │
│           package_name │     NEW     OLD PDIFF   │               NEW               OLD PDIFF   │               NEW               OLD PDIFF   │        NEW        OLD PDIFF   │     NEW     OLD PDIFF   │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│            coq-bignums │   68.86   69.21 -0.51 % │      191968309892      192797833189 -0.43 % │      253804844833      253706717304 +0.04 % │     491188     491208 -0.00 % │     178     117 +52.14 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│         coq-coquelicot │   75.25   75.55 -0.40 % │      206903237145      207139080233 -0.11 % │      250569601069      250489542706 +0.03 % │     656860     656936 -0.01 % │     213      11 +1836.36 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│            coq-unimath │ 2587.98 2596.49 -0.33 % │     7209450127983     7232064769842 -0.31 % │    12998324257464    13000272513353 -0.01 % │    1646336    1606664 +2.47 % │     498     183 +172.13 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│          coq-fourcolor │ 2176.62 2182.22 -0.26 % │     6070685482508     6084881930187 -0.23 % │    11363309620752    11363280659728 +0.00 % │     929336     929396 -0.01 % │       1       0  +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│          coq-fiat-core │  108.45  108.66 -0.19 % │      308604217459      308494144213 +0.04 % │      385521005812      385466281900 +0.01 % │     485388     485408 -0.00 % │     316     164 +92.68 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│              coq-color │  696.77  697.96 -0.17 % │     1948192132587     1950052954432 -0.10 % │     2326340170169     2325776853994 +0.02 % │    1349064    1349072 -0.00 % │      64      99 -35.35 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│   coq-mathcomp-algebra │  166.52  166.71 -0.11 % │      464359362848      464230039703 +0.03 % │      574311060927      574103397017 +0.04 % │     618612     618212 +0.06 % │      82       2 +4000.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│         coq-verdi-raft │ 1324.99 1326.05 -0.08 % │     3695749756533     3700453771672 -0.13 % │     5003409293849     4999742682365 +0.07 % │    1822364    1822320 +0.00 % │       0       1 -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│     coq-mathcomp-field │  230.36  230.52 -0.07 % │      642263147869      641791869869 +0.07 % │      918786148509      918455672348 +0.04 % │     755968     756028 -0.01 % │       0       0  +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│ coq-mathcomp-ssreflect │   39.91   39.92 -0.03 % │      110316651942      110420894050 -0.09 % │      133653992076      133641831351 +0.01 % │     528516     528424 +0.02 % │      11      95 -88.42 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│       coq-fiat-parsers │  635.80  635.77 +0.00 % │     1779604163265     1778226189216 +0.08 % │     2690087377521     2689110914598 +0.04 % │    3233848    3242048 -0.25 % │     303     687 -55.90 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│ coq-mathcomp-character │  187.73  187.62 +0.06 % │      523201171117      522831218908 +0.07 % │      693874651235      693642474707 +0.03 % │    1102052    1102076 -0.00 % │       0       3 -100.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│        coq-lambda-rust │ 1607.52 1605.26 +0.14 % │     4476341245882     4473732241484 +0.06 % │     5960176289444     5956671256929 +0.06 % │    1600892    1600960 -0.00 % │      46       0  +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│  coq-mathcomp-fingroup │   51.58   51.48 +0.19 % │      143868204066      143812401343 +0.04 % │      187200460512      187149658556 +0.03 % │     567912     567956 -0.01 % │       3      31 -90.32 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│       coq-math-classes │  217.20  216.58 +0.29 % │      607335783926      605475000448 +0.31 % │      767899408555      767854834977 +0.01 % │     514508     514608 -0.02 % │      16      13 +23.08 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│             coq-geocoq │ 1945.12 1938.68 +0.33 % │     5429562369827     5408807888599 +0.38 % │     7843634916756     7842568787080 +0.01 % │    1251504    1251328 +0.01 % │      64      74 -13.51 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│  coq-mathcomp-solvable │  189.95  189.21 +0.39 % │      529426729736      528250056759 +0.22 % │      707843040147      707437974516 +0.06 % │     805504     824008 -2.25 % │      20       0  +nan % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│               coq-hott │  334.88  333.44 +0.43 % │      909759629150      906908369036 +0.31 % │     1376923393841     1376708033408 +0.02 % │     477624     477728 -0.02 % │     381      13 +2830.77 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│              coq-verdi │  119.23  118.70 +0.45 % │      331586549004      330289980075 +0.39 % │      421620599572      421457607676 +0.04 % │     572416     572044 +0.07 % │      13       6 +116.67 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│ coq-fiat-crypto-legacy │ 6222.54 6193.90 +0.46 % │    17326639477968    17246247459562 +0.47 % │    28139223547976    28078580067964 +0.22 % │    2656732    2655140 +0.06 % │     965    1057 -8.70 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│               coq-corn │ 1460.14 1453.09 +0.49 % │     4076184584781     4056581530403 +0.48 % │     6019213116317     6015970017752 +0.05 % │     895144     894520 +0.07 % │       1       8 -87.50 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│        coq-fiat-crypto │ 3800.43 3781.96 +0.49 % │    10557468110837    10506813665219 +0.48 % │    16516502608108    16516776011500 -0.00 % │    2548268    2545720 +0.10 % │    1105     951 +16.19 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│ coq-mathcomp-odd-order │ 1469.30 1460.15 +0.63 % │     4094294416661     4070259518172 +0.59 % │     6976284574292     6974665221277 +0.02 % │    1276928    1267256 +0.76 % │      18      18 +0.00 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│           coq-compcert │  964.98  958.29 +0.70 % │     2687530577487     2671683380589 +0.59 % │     3990389575081     3989535853160 +0.02 % │    1032756    1032572 +0.02 % │     165     117 +41.03 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│              coq-flocq │  463.63  459.78 +0.84 % │     1288959976538     1279617002291 +0.73 % │     1777102782973     1776818758973 +0.02 % │    1293136    1293856 -0.06 % │     271      92 +194.57 % │
├────────────────────────┼─────────────────────────┼─────────────────────────────────────────────┼─────────────────────────────────────────────┼───────────────────────────────┼─────────────────────────┤
│             coq-sf-plf │   43.87   43.43 +1.01 % │      122277790575      122594271098 -0.26 % │      158378732073      158371764600 +0.00 % │     481408     488828 -1.52 % │      30       0  +nan % │
└────────────────────────┴─────────────────────────┴─────────────────────────────────────────────┴─────────────────────────────────────────────┴───────────────────────────────┴─────────────────────────┘
~~~

SkySkimmer commented on kernel/environ.mli:
```diff
@@ -46,8 +46,15 @@ type constant_key = Opaqueproof.opaque constant_body * (link_info ref * key)
+  env_modtypes : module_type_body MPmap.t;
+}
+
+type globals = private globals_internal
```
Made private to try to make @ppedrot happier

ppedrot commented:
:goberserk:

ejgallego commented:
> :goberserk: 

:goat: don't consider my question as a blocker for merging; however I am still very curious, rather than an emoji I'd love to see:

- an invariant that is broken [at the ML level] by removing the private from the API

thanks!

ppedrot commented:
> an invariant that is broken [at the ML level] by removing the private from the API

Here is one: the only thing you can observe given an environment is the optional presence of a binding [which is a bug if you take @SkySkimmer's specification literally], and its value.

(And I swear to myself this is the last time I comment in this kind of thread, I am too tired to repeat myself. You all want to get a LISPy model for Coq where you can poke at everything with a stick and get random errors by staring a bit too intensely at a line of code, fine with me. It's obvious that we have so little technical debt in the codebase that it cannot hurt, right?)

(More seriously, I already proposed *several* alternative solutions that would content everybody, but it seems that API design is a memoryless process.)

ejgallego commented:
> the only thing you can observe given an environment is the optional presence of a binding [which is a bug if you take @SkySkimmer's specification literally], and its value.

I am sorry but I don't understand how that is violated, so far it seems to be case.

mattam82 commented:
I don’t get what you are saying about the broken invariant @ppedrot, can you be more explicit/detailed?

ppedrot commented:
A trivial property you can now observe is the order of the keys in the map. This depends on its implementation, which has already changed in the past (from binary trees to hashmaps). In particular it is sensitive to the particular hash function we're tied to (and likely to be different between architectures).

A more involved example is the branch I have started writing to get rid of the canonical / user dialectics for kernel names. In the branch aliasing is explicit via indirections, which breaks any user depending on the actual implementation of the global environment.

And also, I already gave all of these arguments in recurring flamewars over this topic. I am considering writing some wiki entry on the topic, or buy a batch of *Memorex* pills for the dev team.

ppedrot commented:
To add up on a discussion we had at the CUDW, I am also hijacking this PR to ask @ejgallego semi-seriously: why don't you just compile Coq without mli files? It's somewhat trivial to implement and even likely easier with dune, and external users can access whatever they fancy. In the meantime, people trying to keep the system sane are happy, so it's a win-win situation.

ejgallego commented:
> A trivial property you can now observe is the order of the keys in the map.

Umm, you can observe that already, can't you? What can you break with that? I am not sure how this is an invariant. What I am asking is very easy:

- a) assume we have the situation we had in every Coq < 8.9 [globals a record]
- b) then provide some ML code doing something bad.

You are claiming that this is possible, I don't see how; so I am really curious to see what I am missing.

> which breaks any user depending on the actual implementation of the global environment.

This could indeed create an invariant, in this case we would have to provide a view [and I'd be OK with that], but note that this is not the case now.

> I am also hijacking this PR to ask @ejgallego semi-seriously

@ppedrot contrary to what you seem to believe I am having what I regard a technical discussion, I don't need to be reminded as I do still remind I asked you for a concrete example of a broken invariant a so far it has not shown up, what it does show up is use cases for accessing the environment as it is a very reasonable use case.

[we could continue your joke by instead forcing every type to appear in a `.mli` file to be abstract, guess which world would be easy to live]

mattam82 commented:
I understand your point, what would get exposed is not the right abstraction for the env. However I don’t think we lose much by having a private version of it if it helps serialization and does not endanger the kernel. I can’t see a use for updating a globals_internal which is not taken as input by anything.

Relatedly, this is one place where metacoq differs by using a list of declarations ordered by dependencies (mixing constants and inductives) to represent the global environment and we’re considering switching to a similar map abstraction. The immediate issue I have with that is that it is obviously not using the dependency ordering of introductions in the environment. Unless I’m mistaken, there is no way to get that information back from an environ. One must introduce the notion of dependencies between globals, and assume a global universe structure (as otherwise you might have « silent » dependencies between constant declarations). It seems to me (as advocated yesterday by @cyrilcohen) that the really useful abstraction of the env is just this and that’s what people should program against.

ejgallego commented:
> I can’t see a use for updating a globals_internal which is not taken as input by anything.

That'd be mainly out of convenience for clients doing closure of dependency graphs; otherwise they need to clone their own, admittedly that's a very narrow scenario, but if that doesn't cause a problem why not tho?

gares commented:
> why don't you just compile Coq without mli files?

Is it possible? Sometimes exposing symbols is useful for debugging (eg setting breakpoints) so this proposal is not absurd.

ppedrot commented:
@ejgallego I said "semi-seriously" seriously, that is, I was not joking. As noted by @gares debugging is also a natural client for this feature, and SerAPI is strongly comparable to a debugging library for Coq.

ejgallego commented:
> @ejgallego I said "semi-seriously" seriously, that is, I was not joking. As noted by @gares debugging is also a natural client for this feature, and SerAPI is strongly comparable to a debugging library for Coq.

I dunno, I am not sure I'd like to mess with that.

From my point of view SerAPI should just reify the public-facing API, I'd like to keep things simple.

ejgallego commented:
What's holding this? Note that #8228 may provide a preferred solution?

But indeed I'm not sure I feel comfortable with this PR, the description says "If you have access to a kernel name you also should have the environment in which it is defined"

I am not sure this is always going to be case; indeed we should have a discussion about what the use cases for the kernel environment are, different views on this have caused much turnoil in different PRs.

To me, an environment is precisely something you can query as a client; I have a hard time finding a functional language that was able to work with an opaque env.

Note that I don't make a distinction of the global vs the named env for example. Imagine you are writing a tactic, or some search. You either can query the environment in a proper way, or you are forced to duplicate the environment somewhere in the upper layers, and indeed this is mostly done in Coq, and IMVHO I have problems justifying that duplication, that without taking into account all the bugs we have due to the duplication of environment data in form of libobjects .

So I dunno, it seems clear to me that read-only access to the environment is much needed, a different story is about construction tho, but even in that case it is not hard to imagine applications [incremental type-checking for example] that would make the trusted core quite fat due to a construction restriction.

Anyways, I hope we can make advance on the PR in any direction.

SkySkimmer commented:
Changed: reverted exporting globals, added `mem_constant/mind` to environ interface
Bench waiting to start n°754

SkySkimmer commented:
Looks like I missed a mem_constant, let's try again.

SkySkimmer commented:
Looks like it works

SkySkimmer commented:
ping @ppedrot
